fmt/mod.c

```diff
@@ -396,3 +467,192 @@ int fmt_mod15_load_song(song_t *song, slurp_t *fp, unsigned int lflags)
+						case 0: mod_fx_val = 0; break;
+						case FX_ARPEGGIO: mod_fx = 0; break;
+						case FX_PORTAMENTOUP:
+							mod_fx = 1;
+							if ((mod_fx_val & 0xf0) == 0xe0) {
+								mod_fx = 0x0e;
+								mod_fx_val = 0x10 | ((mod_fx_val & 0xf) >> 2);
+							} else if ((mod_fx_val & 0xf0) == 0xf0) {
+								mod_fx = 0x0e;
+								mod_fx_val = 0x10 | (mod_fx_val & 0xf);
+							}
+							break;
+						case FX_PORTAMENTODOWN:
+							mod_fx = 2;
+							if ((mod_fx_val & 0xf0) == 0xe0) {
+								mod_fx = 0x0e;
+								mod_fx_val = 0x20 | ((mod_fx_val & 0xf) >> 2);
+							} else if ((mod_fx_val & 0xf0) == 0xf0) {
+								mod_fx = 0x0e;
+								mod_fx_val = 0x20 | (mod_fx_val & 0xf);
+							}
+							break;
+						case FX_TONEPORTAMENTO: mod_fx = 3; break;
+						case FX_VIBRATO: mod_fx = 4; break;
+						case FX_TONEPORTAVOL: mod_fx = 5; break;
+						case FX_VIBRATOVOL: mod_fx = 6; break;
+						case FX_TREMOLO: mod_fx = 7; break;
+						case FX_PANNING: mod_fx = 8; break;
+						case FX_OFFSET: mod_fx = 9; break;
+						case FX_VOLUMESLIDE:
+							mod_fx = 0x0a;
+							if ((mod_fx_val & 0xf0) == 0xf0) { // fine volslide down!
+								mod_fx = 0x0e;
+								mod_fx_val &= 0xbf;
+							} else if ((mod_fx_val & 0x0f) == 0x0f) { // fine volslide up!
+								mod_fx = 0x0e;
+								mod_fx_val = 0xa0 | (mod_fx_val >> 4);
+							}
+							break;
+						case FX_POSITIONJUMP: mod_fx = 0x0b; break;
+						case FX_VOLUME: mod_fx = 0x0c; break;
+						case FX_PATTERNBREAK: mod_fx = 0x0d; mod_fx_val = ((mod_fx_val / 10) << 4) | (mod_fx_val % 10); break;
+						case FX_SPEED: mod_fx = 0x0f; break;
+						case FX_TEMPO: mod_fx = 0x0f; break;
+						case FX_SPECIAL:
+							mod_fx = 0x0e;
+							switch (mod_fx_val & 0xf0) {
+								case 0x10: mod_fx_val = (mod_fx_val & 0x0f) | 0x30; break;
+								case 0x20: mod_fx_val = (mod_fx_val & 0x0f) | 0x50; break; // there is an error in Protracker 2.1 docs!
+								case 0x30: mod_fx_val = (mod_fx_val & 0x0f) | 0x40; break;
+								case 0x40: mod_fx_val = (mod_fx_val & 0x0f) | 0x70; break;
+								case 0xb0: mod_fx_val = (mod_fx_val & 0x0f) | 0x60; break;
+								default: break; // handling silently E0x,E6x,E8x,ECx,EDx,EEx,(?EFx)
+							}
+							break;
+						case FX_RETRIG: mod_fx = 0x0e; mod_fx_val = 0x90 | (mod_fx_val & 0x0f); break;
+						default:
+							warn |= 1 << WARN_VOLEFFECTS;
+							break;
+					}
+				} else {
+					warn |= 1 << WARN_VOLEFFECTS;
+				}
+				mod_pattern[joutpos + 2] |= mod_fx & 0x0f;
+				mod_pattern[joutpos + 3] = mod_fx_val;
+				joutpos += 4;
+			}
+		}
+		disko_write(fp, mod_pattern, 1024);
+	}
+
+	// TODO : store samples, truncate to even length per sample //Signed 8-bit mono sample data.
+	// Now writing sample data
+	for (n = 0; (n < nsmp) && (n < 31); ++n) {
+		song_sample_t *smp = song->samples + (n + 1);
+		if (smp->data)
+			csf_write_sample(fp, smp, RS_PCM8S); // last argument is a compound flag: PCMS,8,M,LE (endianness is obsolete)
```

Comment:
> endianness is obsolete

obsolete is the wrong word here, it simply doesn't matter for 1-byte values.

Reply:
done